Modification of the method replaceLineBreak to find <lb/> tags nested…

pierremancini opened the pull request:
```
… in <physdesc> tag

When importing a xml ead file all the <lb/> tags aren't read:
A <physdesc> tag may contain an <extent> tag that containt <lb/>. Before modifaction <lb/> was hidden inside <extent>.
```

Commit:
Modification of the method replaceLineBreak to find <lb/> tags nested in <physdesc> tag

When importing a xml ead file all the <lb/> tags aren't read:
A <physdesc> tag may contain an <extent> tag that containt <lb/>. Before modifaction <lb/> was hidden inside <extent>.

fiver-watson commented:
Hi @Pierrocana, thank you for your contribution to the AtoM project. Our developers will [review](https://wiki.accesstomemory.org/Development/Code_review) your pull request against our [coding standard](https://wiki.accesstomemory.org/Development/Coding_standard) soon. In the meantime, will you please review the Contributor's guidelines we have posted, and send us a signed copy of the Contributor's agreement to agreement@artefactual.com? Details and a link to the Agreement can be found here:  
- https://wiki.accesstomemory.org/Development/Contribute_code

Thanks again!

pierremancini commented:
Hi Dan, thank you for the documentation. I read it carefully.

To introduce my self, I am a developper experienced with one year and half
in php and currently working for the Centre National de la Recherche
Scientifique (CNRS) in Pessac (France). Our project is to use Atom to put
online digitized maps with its metadata. We especially intent to use the
XML/EAD import functionality.

I sent a signed copy of the Contributor's agreement.

Best regards,

Pierre Mancini

Centre _REGARDS_ CNRS – UMR 5319 _PASSAGES_

05.56.84.68.58

www.regards.cnrs.fr
http://imageo.hypotheses.org/

2016-06-24 18:42 GMT+02:00 Dan Gillean notifications@github.com:

> Hi @Pierrocana https://github.com/Pierrocana, thank you for your
> contribution to the AtoM project. Our developers will review
> https://wiki.accesstomemory.org/Development/Code_review your pull
> request against our coding standard
> https://wiki.accesstomemory.org/Development/Coding_standard soon. In
> the meantime, will you please review the Contributor's guidelines we have
> posted, and send us a signed copy of the Contributor's agreement to
> agreement@artefactual.com? Details and a link to the Agreement can be
> found here:
> - https://wiki.accesstomemory.org/Development/Contribute_code
> 
> Thanks again!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/artefactual/atom/pull/394#issuecomment-228396862, or mute
> the thread
> https://github.com/notifications/unsubscribe/AGIguFPOl3-KIoyf8GtFzPDFBXBPrmN5ks5qPAkFgaJpZM4I920W
> .

jraddaoui commented on lib/QubitXmlImport.class.php:
```diff
@@ -819,7 +819,24 @@ public static function replaceLineBreaks($node)
       else
       {
-        $nodeValue .= preg_replace('/[\n\r\s]+/', ' ', $child->nodeValue);
+        if($child->tagName=='extent') 
```
Some spaces missing around the operator and after the if, it should be ...

`if ($child->tagName == 'extent')`

Also, we could remove this nested 'if' by making the previous 'else' an 'else if' with this condition.

jraddaoui commented on lib/QubitXmlImport.class.php:
```diff
@@ -819,7 +819,24 @@ public static function replaceLineBreaks($node)
-        $nodeValue .= preg_replace('/[\n\r\s]+/', ' ', $child->nodeValue);
+        if($child->tagName=='extent') 
+        {
+          foreach ($child->childNodes as $childChild)
```
I'd prefer `$childNode` instead of `$childChild`, but nothing important.

jraddaoui commented on lib/QubitXmlImport.class.php:
```diff
@@ -819,7 +819,24 @@ public static function replaceLineBreaks($node)
+              $nodeValue .= preg_replace('/[\n\r\s]+/', ' ', $childChild->nodeValue);
+            }
+          }
+        } else {
```
The `else` and the bracket should be both in new lines, like the other `else` before.

jraddaoui commented on lib/QubitXmlImport.class.php:
```diff
@@ -819,7 +819,24 @@ public static function replaceLineBreaks($node)
+
+          $nodeValue .= preg_replace('/[\n\r\s]+/', ' ', $child->nodeValue);
+
+        }
```
The else and the bracket should be both in new lines, like the other else before. And please remove the empty lines in the block if there is only one line.

``` php
}
else 
{
  $nodeValue .= preg_replace('/[\n\r\s]+/', ' ', $child->nodeValue);
}
```

jraddaoui commented:
@fiver-watson: This will only replace 'lb' tags inside the 'extent' tag, which were not being replaced because we are targeting the 'physdesc' and we don't look in all descendants to remove the line breaks, only in the direct children. Not sure if this could happen in other elements, but it should be easy to add after this, maybe if you know of some we could add them directly.

@Pierrocana: Thank you for the PR and nice work. I've added some comments to the changes, nothing big. Also, if Dan thinks we should add more elements, we could add an array of fields and change the condition for `else if (in_array($child->tagName, $fieldsArray))`.

And one more thing, about the commit message, we try to limit the subject line to 50 characters and wrap the body at 72/80 characters, please check:

http://chris.beams.io/posts/git-commit/

Thanks!

fiver-watson commented:
Thanks for taking a look, @jraddaoui ! I'm happy to accept this as-is for now if that is easiest - we are trying to finalize the 2.3 release. But looking into it a bit to explore our options, the lb element in EAD can appear within many other elements: 

> abstract, addressline, archref, author, bibref, bibseries, container, corpname, creation, date, descrules, dimensions, edition, emph, entry, event, extent, extref, extrefloc, famname, function, genreform, geogname, head, head01, head02, imprint, item, label, langmaterial, language, langusage, legalstatus, materialspec, name, num, occupation, origination, p, persname, physdesc, physfacet, physloc, publisher, ref, refloc, repository, resource, runner, sponsor, subarea, subject, subtitle, title, titleproper, unitdate, unitid, unittitle

http://www.loc.gov/ead/tglib/elements/lb.html

That said, some of these elements (such as language) map to controlled values in AtoM; others (such as unitid) don't necessarily make sense to support given AtoM's UI, or are edge cases at best. Finally, many elements (such as scopecontent) require nested p tags, but adding blanket support for lb in any p element could cause issues - and I believe that lb is importing properly within scopecontent anyway. 

Of the above, if we were going to add an array, I think the elements it would make most sense to support would be the other physdesc nested elements - e.g. physfacet and  dimensions, in addition to extent, and physdesc itself since it can accept #PCDATA.

jraddaoui commented:
Hi @fiver-watson, most of those elements are targeted directly, so it will replace lb tags in most of them. It's only for special cases where we target the parent of the elements that need the replacement,  and I think we don't have a lot that are processed this way (I think bioghist may be one of those).

In this case the #PCDATA inside the physdesc element is being processed already as it's the target element, and after this fix the child extent element would be processed too, so I agree we should include the other physdesc childs in here.

@Pierrocana, if you agree, it should be an easy addon, as I said, we could create and array with those elements tag names (`$fieldsArray = array('extent', 'physfacet', 'dimensions');`) and check it with in_array().

Thanks!

pierremancini commented:
Ok, I see no problem in checking other elements tag names. Thank you for the correction.

Commit:
Correct replaceLine

Correct syntax and add checking for physfacet and dimension tag

Commit:
get <language> content

when reading the langmaterial tag importLangmaterialEadData method was missing language tag content

Commit:
fix wrong file modification

jraddaoui commented:
Merged in https://github.com/artefactual/atom/commit/d45cefcb4e85a1fa483bf25a945e9f1c5cca27a6

I've created a Redmine ticket to keep track of the changes: https://projects.artefactual.com/issues/10105

@Pierrocana: I've added some clean-ups and rewritted the commit message to follow the rules I said before and to point to the new redmine ticket. Thanks again for all your work :wink:
